Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm

```diff
@@ -77,28 +78,76 @@
+
+            bool isWithinPageWidth = isInRange(documentSpacePoint.x(), pageLayoutBounds.x(), pageLayoutBounds.maxX());
+            bool isWithinPageHeight = isInRange(documentSpacePoint.y(), pageLayoutBounds.y(), pageLayoutBounds.maxY());
+
+            if (isAbovePage && isWithinPageWidth)
+                pointsToCompare.append({ documentSpacePoint.x(), pageLayoutBounds.y() });
+            else if (isRightOfPage && isWithinPageHeight)
+                pointsToCompare.append({ pageLayoutBounds.x(), documentSpacePoint.y() });
+            else if (isBelowPage && isWithinPageWidth)
+                pointsToCompare.append({ documentSpacePoint.x(), pageLayoutBounds.maxY() });
+            else if (isLeftOfPage && isWithinPageHeight)
+                pointsToCompare.append({ pageLayoutBounds.x(), documentSpacePoint.y() });
+
+            auto distancesToPoints = WTF::map(pointsToCompare, [documentSpacePoint](FloatPoint point) {
+                return euclidianDistance(point, documentSpacePoint);
+            });
 
-            auto currentPageBounds = layoutBoundsForPageAtIndex(index);
-            if (documentSpacePoint.y() < currentPageBounds.maxY()) {
+            return *std::min_element(distancesToPoints.begin(), distancesToPoints.end());
+        };
 
-                auto pairedPageIndex =  [index, this]() {
-                    if (index % pagesPerRow())
-                        return index - 1;
+        for (PageIndex index = 0; index < pageCount; index += pagesPerRow()) {
+            auto [leftPageLayoutBounds, rightPageLayoutBounds]= layoutBoundsForPairedPages(index);
+
+            if (documentSpacePoint.y() < leftPageLayoutBounds.maxY() || (rightPageLayoutBounds && documentSpacePoint.y() < rightPageLayoutBounds->maxY())) {
+
+                if (leftPageLayoutBounds.contains(documentSpacePoint) || !rightPageLayoutBounds)
+                    return index;
+
+                if (rightPageLayoutBounds->contains(documentSpacePoint))
                     return index + 1;
-                }();
 
-                auto pairedPageMaxX= layoutBoundsForPageAtIndex(pairedPageIndex).maxX();
-                auto currentPageMaxX = currentPageBounds.maxX();
-                if (currentPageMaxX < pairedPageMaxX)
-                    return documentSpacePoint.x() < currentPageMaxX ? index : pairedPageIndex;
-                return documentSpacePoint.x() < pairedPageMaxX ? pairedPageIndex : index;
+                if (distanceToPage(leftPageLayoutBounds) < distanceToPage(rightPageLayoutBounds.value()))
+                    return index;
+
+                return index + 1;
             }
         }
-        break;
+
+        auto lastPageIndex = pageCount - 1;
+        if (isLeftPageIndex(lastPageIndex))
+            return lastPageIndex;
+
+        auto [leftPageLayoutBounds, rightPageLayoutBounds] = layoutBoundsForPairedPages(lastPageIndex);
+        if (distanceToPage(leftPageLayoutBounds) < distanceToPage(rightPageLayoutBounds.value()))
```

Comment:
Same thing here, the `lastPageIndex` will result in a `nullopt` value for `rightPageLayoutBounds`, right

Reply:
Similar to above, with the conditional of the "is the last page in the document a page on the left side," but you are right that technically we don't explicitly check that the result from `layoutBoundsForPairedPages` is not empty. I added an `ASSERT` to indicate that it would be very weird if it was. Also added an additional condition just to make sure also